storage/engine/mvcc.go

```diff
@@ -432,53 +444,71 @@ func MVCCPutProto(engine Engine, ms *MVCCStats, key proto.Key, timestamp proto.T
 // keyA_Timestamp_0 : value of version_0
 // keyB : MVCCMetadata of keyB
 // ...
-func MVCCGet(engine Engine, key proto.Key, timestamp proto.Timestamp, txn *proto.Transaction) (*proto.Value, error) {
+func MVCCGet(engine Engine, key proto.Key, timestamp proto.Timestamp,
+	txn *proto.Transaction) (*proto.Value, error) {
 	if len(key) == 0 {
 		return nil, emptyKeyError()
 	}
 
 	// Create a function which scans for the first key between next and end keys.
-	earlier := func(engine Engine, start, end proto.EncodedKey) (proto.RawKeyValue, error) {
+	getValue := func(engine Engine, start, end proto.EncodedKey,
+		msg gogoproto.Message) (proto.EncodedKey, error) {
 		iter := engine.NewIterator()
 		defer iter.Close()
 		iter.Seek(start)
-		if iter.Valid() && bytes.Compare(iter.Key(), end) < 0 {
-			return proto.RawKeyValue{Key: iter.Key(), Value: iter.Value()}, nil
+		if !iter.Valid() {
+			return nil, iter.Error()
 		}
-		return proto.RawKeyValue{}, iter.Error()
+		key := iter.Key()
+		if bytes.Compare(key, end) >= 0 {
+			return nil, iter.Error()
+		}
+		return key, iter.ValueProto(msg)
 	}
 
-	metaKey := MVCCEncodeKey(key)
-	data, err := engine.Get(metaKey)
-	if err != nil || data == nil {
+	buf := getBufferPool.Get().(*getBuffer)
+
+	metaKey := mvccEncodeKey(buf.key[0:0], key)
+	ok, _, _, err := engine.GetProto(metaKey, &buf.meta)
+	if err != nil || !ok {
+		getBufferPool.Put(buf)
 		return nil, err
 	}
 
-	return mvccGetInternal(engine, key, proto.RawKeyValue{Key: metaKey, Value: data}, timestamp, txn, earlier)
+	value, err := mvccGetInternal(engine, key, metaKey, timestamp, txn, getValue, buf)
+	var rvalue *proto.Value
+	if err == nil && value != nil {
+		rvalue = &proto.Value{}
+		*rvalue = *value
+	}
+	getBufferPool.Put(buf)
+	return rvalue, err
 }
 
 // getEarlierFunc fetches an earlier version of a key starting at
 // start and ending at end. Returns the value as a byte slice, the
 // timestamp of the earlier version, a boolean indicating whether a
 // version value or metadata was found, and error, if applicable.
-type getEarlierFunc func(engine Engine, start, end proto.EncodedKey) (proto.RawKeyValue, error)
+type getValueFunc func(engine Engine, start, end proto.EncodedKey,
```

Comment:
I was confused by the previous code that `getEarlierFunc` was being used to retrieve both the metadata and the values. The `getValueFunc` is only used to retrieve values. This seems cleaner to me.